lib/utils/suppressionsService.mjs

```diff
@@ -0,0 +1,252 @@
+/*
+ * This file is based on ESLint's suppressions-service.js
+ * https://github.com/eslint/eslint/blob/v9.26.0/lib/services/suppressions-service.js
+ *
+ * Copyright OpenJS Foundation and other contributors, https://openjsf.org/
+ * Released under the MIT License:
+ * https://github.com/eslint/eslint/blob/main/LICENSE
+ */
+import fs from 'node:fs';
+import path from 'node:path';
+
+import getRelativePath from './getRelativePath.mjs';
+import isPathNotFoundError from './isPathNotFoundError.mjs';
+
+/** @import {LintResult, Warning, SuppressedViolations} from 'stylelint' */
+
+/**
+ * Manages the suppressed violations.
+ */
+export class SuppressionsService {
+	filePath = '';
+	cwd = '';
+
+	/**
+	 * Creates a new instance of SuppressionsService.
+	 * @param {Object} options The options.
+	 * @param {string} options.filePath The path to the suppressions file.
+	 * @param {string} options.cwd The current working directory.
+	 */
+	constructor({ filePath, cwd }) {
+		this.filePath = filePath;
+		this.cwd = cwd;
+	}
+
+	/**
+	 * Updates the suppressions file based on the current violations and the provided rules.
+	 * If no rules are provided, all violations are suppressed.
+	 * @param {LintResult[] | undefined} results The lint results.
+	 * @param {string[] | undefined} rules The rules to suppress.
+	 * @returns {Promise<void>}
+	 */
+	async suppress(results, rules) {
+		const suppressions = await this.load();
+
+		if (results === undefined) return;
+
+		for (const result of results) {
+			const source = result.source;
+
+			if (!source) continue;
+
+			const relativePath = path.isAbsolute(source) ? getRelativePath(this.cwd, source) : source;
+
+			const violationsByRule = SuppressionsService.countViolationsByRule(result.warnings);
+
+			for (const rule of Object.keys(violationsByRule)) {
+				if (rules && !rules.includes(rule)) continue;
+
+				suppressions[relativePath] ??= {};
+
+				if (!violationsByRule[rule]) continue;
+
+				suppressions[relativePath][rule] = violationsByRule[rule];
+			}
+		}
+
+		return this.save(suppressions);
+	}
+
+	/**
+	 * Removes old, unused suppressions for violations that do not occur anymore.
+	 * @param {LintResult[]} results The lint results.
+	 * @returns {Promise<void>} No return value.
+	 */
+	async prune(results) {
+		const suppressions = await this.load();
+		const { unused } = this.applySuppressions(results, suppressions);
+
+		for (const file in unused) {
```

Comment:
[question] I'm not sure about the background on the ESLint side, I'm concerned why `for...in` is heavily used here. In modern JavaScript, I understand `for...of` is better than `for...in` in many cases.

MDN also says in the [`for...in`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...in) page:

> Many JavaScript style guides and linters recommend against the use of `for...in`, because it iterates over the entire prototype chain which is rarely what one wants, and may be a confusion with the more widely-used `for...of` loop.

Is there any reason?

Reply:
I use `for...in` only where we’re enumerating the keys of plain objects. Switching to `for...of` would require calling `Object.keys()` each time to create an array first, adding unnecessary overhead.

Reply:
If `Map` is available instead of `Object`, `for...of` is easier to use. What about `Map` instead?

Ref: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map#objects_vs._maps